File: 程式設計：資料結構答案檔/C#/301.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

class Player
{
    public int Id { get; set; }
    public string Name { get; set; }
    public int Level { get; set; }
    public int Score { get; set; }
}

class Program
{
    static void Main()
    {
        var filePath = "read.csv";
        var rawData = ReadCsvFile(filePath);

        if (!rawData.Any())
        {
            Console.WriteLine("Error reading CSV file or file is empty.");
            return;
        }

        List<Player> players = ParsePlayerData(rawData);

        if (!int.TryParse(Console.ReadLine(), out int n))
        {
            Console.WriteLine("Invalid input");
            return;
        }

        DisplaySortedPlayerData(players, n);
    }

    static List<string[]> ReadCsvFile(string filePath)
    {
        var data = new List<string[]>();
        try
        {
            using (var reader = new StreamReader(File.OpenRead(filePath)))
            {
                while (!reader.EndOfStream)
                {
                    var line = reader.ReadLine();
                    var values = line.Split(',');
                    data.Add(values);
                }
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error reading CSV file: {ex.Message}");
        }
        return data;
    }

    static List<Player> ParsePlayerData(List<string[]> rawData)
    {
        return rawData.Skip(1)
            .Select(data => new Player()
            {
                Id = int.Parse(data[0]),
                Name = data[1],
                Level = int.Parse(data[2]),
                Score = int.Parse(data[3])
            }).ToList();
    }

    static void DisplaySortedPlayerData(List<Player> players, int n)
    {
        if (n > players.Count)
        {
            Console.WriteLine("Exceeds data upper limit");
            return;
        }

        var sortedPlayers = players.OrderByDescending(p => p.Score).ThenByDescending(p => p.Id).Take(n);

        int rank = 1;
        foreach (var player in sortedPlayers)
        {
            Console.WriteLine($"{rank} | {player.Id} | {player.Name} | {player.Level} | {player.Score}");
            rank++;
        }
    }
}

File: 程式設計：資料結構答案檔/Python/301.py
import csv

class Player:
    def __init__(self, Id, Name, Level, Score):
        self.Id = int(Id)
        self.Name = Name
        self.Level = int(Level)
        self.Score = int(Score)

    def __repr__(self):
        return f"{self.Id} | {self.Name} | {self.Level} | {self.Score}"

def read_csv_file(file_path):
    players = []
    with open(file_path, newline='') as csvfile:
        reader = csv.reader(csvfile)
        next(reader, None)  # Skip the header
        for row in reader:
            player = Player(*row)
            players.append(player)
    return players

def sort_players(players):
    return sorted(players, key=lambda x: (-x.Score, -x.Id))

def display_sorted_player_data(players, n):
    if n > len(players):
        print("Exceeds data upper limit")
        return
    for index, player in enumerate(players[:n], start=1):
        print(f"{index} | {player}")

def main():
    file_path = "read.csv"
    players = read_csv_file(file_path)
    sorted_players = sort_players(players)
    
    try:
        n = int(input())
        display_sorted_player_data(sorted_players, n)
    except ValueError:
        print("Please enter a valid number.")

if __name__ == "__main__":
    main()
